// ng2-flatpickr/esm2015/src/ng2-flatpickr.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Ng2FlatpickrComponent } from './ng2-flatpickr.component';
import { Ng2FlatpickrDirective } from './ng2-flatpickr.directive';
import 'flatpickr';
import * as ɵngcc0 from '@angular/core';
let Ng2FlatpickrModule = class Ng2FlatpickrModule {
};
Ng2FlatpickrModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Ng2FlatpickrModule });
Ng2FlatpickrModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Ng2FlatpickrModule_Factory(t) { return new (t || Ng2FlatpickrModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Ng2FlatpickrModule, { declarations: function () { return [Ng2FlatpickrComponent,
        Ng2FlatpickrDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [Ng2FlatpickrComponent,
        Ng2FlatpickrDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Ng2FlatpickrModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [
                    Ng2FlatpickrComponent,
                    Ng2FlatpickrDirective
                ],
                exports: [
                    Ng2FlatpickrComponent,
                    Ng2FlatpickrDirective
                ]
            }]
    }], null, null); })();
export { Ng2FlatpickrModule };
