// ngx-bar-rating/esm2015/lib/bar-rating.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { BarRating } from './bar-rating';
import { BarRatingPipe } from './bar-rating.pipe';
import * as ɵngcc0 from '@angular/core';
export class BarRatingModule {
}
BarRatingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: BarRatingModule });
BarRatingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function BarRatingModule_Factory(t) { return new (t || BarRatingModule)(); }, imports: [[
            CommonModule,
            FormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BarRatingModule, { declarations: function () { return [BarRating, BarRatingPipe]; }, imports: function () { return [CommonModule,
        FormsModule]; }, exports: function () { return [BarRating]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BarRatingModule, [{
        type: NgModule,
        args: [{
                declarations: [BarRating, BarRatingPipe],
                imports: [
                    CommonModule,
                    FormsModule
                ],
                exports: [BarRating]
            }]
    }], null, null); })();
