// ng-wizard/esm2015/lib/core/ng-wizard-step-content.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ViewContainerRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
let NgWizardStepContentDirective = class NgWizardStepContentDirective {
    constructor(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
    }
};
NgWizardStepContentDirective.ɵfac = function NgWizardStepContentDirective_Factory(t) { return new (t || NgWizardStepContentDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
NgWizardStepContentDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NgWizardStepContentDirective, selectors: [["", "ngWizardStepContent", ""]] });
NgWizardStepContentDirective.ctorParameters = () => [
    { type: ViewContainerRef }
];
NgWizardStepContentDirective = __decorate([ __metadata("design:paramtypes", [ViewContainerRef])
], NgWizardStepContentDirective);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgWizardStepContentDirective, [{
        type: Directive,
        args: [{
                selector: '[ngWizardStepContent]'
            }]
    }], function () { return [{ type: ɵngcc0.ViewContainerRef }]; }, null); })();
export { NgWizardStepContentDirective };
