// @ng-select/ng-select/esm2015/ng-select-ng-select.js
/**
 * @fileoverview added by tsickle
 * Generated from: ng-select-ng-select.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { NgSelectComponent, SELECTION_MODEL_FACTORY, NgSelectModule, NgSelectConfig } from './public-api';
export { ConsoleService as ɵs } from './lib/console.service';
export { NgDropdownPanelComponent as ɵq } from './lib/ng-dropdown-panel.component';
export { NgDropdownPanelService as ɵd } from './lib/ng-dropdown-panel.service';
export { NgOptionComponent as ɵr } from './lib/ng-option.component';
export { NgFooterTemplateDirective as ɵk, NgHeaderTemplateDirective as ɵj, NgItemLabelDirective as ɵe, NgLabelTemplateDirective as ɵh, NgLoadingSpinnerTemplateDirective as ɵp, NgLoadingTextTemplateDirective as ɵn, NgMultiLabelTemplateDirective as ɵi, NgNotFoundTemplateDirective as ɵl, NgOptgroupTemplateDirective as ɵg, NgOptionTemplateDirective as ɵf, NgTagTemplateDirective as ɵo, NgTypeToSearchTemplateDirective as ɵm } from './lib/ng-templates.directive';
export { DefaultSelectionModel as ɵc, DefaultSelectionModelFactory as ɵb } from './lib/selection-model';
