// ngx-bar-rating/bundles/ngx-bar-rating.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('@angular/forms'),exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-bar-rating', ['@angular/core','@angular/common','@angular/forms','exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.ng.core,global.ng.common,global.ng.forms,global['ngx-bar-rating'] = {}, global.ng.core, global.ng.forms, global.ng.common));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,exports, core, forms, common) { 
function BarRating_div_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵlistener("click", function BarRating_div_2_Template_div_click_0_listener($event) { var unit_r2 = ctx.$implicit; return unit_r2.click($event); })("mouseenter", function BarRating_div_2_Template_div_mouseenter_0_listener() { var unit_r2 = ctx.$implicit; return unit_r2.enter(); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var unit_r2 = ctx.$implicit;
    ɵngcc0.ɵɵclassProp("br-fraction", unit_r2.fraction)("br-selected", unit_r2.selected)("br-active", unit_r2.active);
} }
function BarRating_div_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 4);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵpipe(2, "rateTitle");
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate(ɵngcc0.ɵɵpipeBind2(2, 1, ctx_r1.nextRate, ctx_r1.titles));
} }
'use strict';

    /** This allows support [(ngModel)] and ngControl. */
    var RATING_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        useExisting: core.forwardRef(function () { return BarRating; }),
        multi: true
    };
    /** This allows control required validation. */
    var RATING_VALUE_VALIDATOR = {
        provide: forms.NG_VALIDATORS,
        useExisting: core.forwardRef(function () { return BarRating; }),
        multi: true,
    };
    var BarRating = /** @class */ (function () {
        function BarRating(changeDetectorRef) {
            this.changeDetectorRef = changeDetectorRef;
            this.contexts = [];
            /** Maximal rating that can be given using this widget. */
            this.max = 5;
            /** A flag indicating if rating can be updated. */
            this.readOnly = false;
            /** Set the theme */
            this.theme = 'default';
            /** Show rating title */
            this.showText = false;
            /** Replace rate value with a title */
            this.titles = [];
            /** A flag indicating if rating is required for form validation. */
            this.required = false;
            /**
             * An event fired when a user is hovering over a given rating.
             * Event's payload equals to the rating being hovered over.
             */
            this.hover = new core.EventEmitter();
            /**
             * An event fired when a user stops hovering over a given rating.
             * Event's payload equals to the rating of the last item being hovered over.
             */
            this.leave = new core.EventEmitter();
            /**
             * An event fired when a user selects a new rating.
             * Event's payload equals to the newly selected rating.
             */
            this.rateChange = new core.EventEmitter(true);
            this.onChange = function (_) {
            };
            this.onTouched = function () {
            };
        }
        BarRating.prototype.ngOnChanges = function (changes) {
            if (changes.rate) {
                this.update(this.rate);
            }
        };
        BarRating.prototype.ngOnInit = function () {
            var _this = this;
            this.contexts = Array.from({ length: this.max }, function (context, i) { return ({
                selected: false,
                fraction: false,
                active: false,
                click: function (e) { return _this.handleClick(e, i + 1); },
                enter: function () { return _this.handleEnter(i + 1); }
            }); });
            this.updateState(this.rate);
        };
        BarRating.prototype.update = function (newRate, internalChange) {
            if (internalChange === void 0) { internalChange = true; }
            if (!this.readOnly && !this.disabled && this.rate !== newRate) {
                this.rate = newRate;
                this.rateChange.emit(this.rate);
            }
            if (internalChange) {
                this.onChange(this.rate);
                this.onTouched();
            }
            this.updateState(this.rate);
        };
        /** Reset rate value */
        BarRating.prototype.reset = function () {
            this.leave.emit(this.nextRate);
            this.updateState(this.rate);
        };
        BarRating.prototype.updateState = function (nextValue) {
            var _this = this;
            /** Set rate value as text */
            this.nextRate = nextValue - 1;
            /** Set the rating */
            this.contexts = Array.from({ length: this.max }, function (context, index) { return ({
                selected: index + 1 <= nextValue,
                fraction: (index + 1 === Math.round(nextValue) && nextValue % 1) >= 0.5,
                active: false,
                click: function (e) { return _this.handleClick(e, index); },
                enter: function () { return _this.handleEnter(index); }
            }); });
        };
        BarRating.prototype.handleClick = function (e, value) {
            /** (NOT TESTED) Remove 300ms click delay on touch devices */
            e.preventDefault();
            e.stopPropagation();
            this.update(value + 1);
        };
        BarRating.prototype.handleEnter = function (index) {
            if (!this.disabled && !this.readOnly) {
                /** Add selected class for rating hover */
                this.contexts.map(function (context, i) {
                    context.active = i <= index;
                    context.fraction = false;
                    context.selected = false;
                });
                this.nextRate = index;
                this.hover.emit(index);
            }
        };
        /** This is the initial value set to the component */
        BarRating.prototype.writeValue = function (value) {
            this.update(value, false);
            this.changeDetectorRef.markForCheck();
        };
        BarRating.prototype.validate = function (c) {
            return (this.required && !c.value) ? { required: true } : null;
        };
        BarRating.prototype.registerOnChange = function (fn) {
            this.onChange = fn;
        };
        BarRating.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        BarRating.prototype.setDisabledState = function (isDisabled) {
            this.disabled = isDisabled;
        };
BarRating.ɵfac = function BarRating_Factory(t) { return new (t || BarRating)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
BarRating.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: BarRating, selectors: [["bar-rating"]], inputs: { max: "max", readOnly: "readOnly", theme: "theme", showText: "showText", titles: "titles", required: "required", rate: "rate" }, outputs: { hover: "hover", leave: "leave", rateChange: "rateChange" }, features: [ɵngcc0.ɵɵProvidersFeature([RATING_VALUE_ACCESSOR, RATING_VALUE_VALIDATOR]), ɵngcc0.ɵɵNgOnChangesFeature], decls: 4, vars: 9, consts: [[1, "br-units", 3, "mouseleave"], ["class", "br-unit", 3, "br-fraction", "br-selected", "br-active", "click", "mouseenter", 4, "ngFor", "ngForOf"], ["class", "br-text", 4, "ngIf"], [1, "br-unit", 3, "click", "mouseenter"], [1, "br-text"]], template: function BarRating_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div");
        ɵngcc0.ɵɵelementStart(1, "div", 0);
        ɵngcc0.ɵɵlistener("mouseleave", function BarRating_Template_div_mouseleave_1_listener() { return ctx.reset(); });
        ɵngcc0.ɵɵtemplate(2, BarRating_div_2_Template, 1, 6, "div", 1);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, BarRating_div_3_Template, 3, 4, "div", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMapInterpolate1("br br-", ctx.theme, "");
        ɵngcc0.ɵɵclassProp("br-readonly", ctx.readOnly)("br-disabled", ctx.disabled);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.contexts);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.showText);
    } }, directives: function () { return [ɵngcc1.NgForOf, ɵngcc1.NgIf]; }, pipes: function () { return [BarRatingPipe]; }, styles: ["*[_ngcontent-%COMP%]{box-sizing:border-box}.br[_ngcontent-%COMP%]{margin:15px 0;position:relative}.br-units[_ngcontent-%COMP%]{display:flex;flex-wrap:nowrap}.br-unit[_ngcontent-%COMP%]{-webkit-font-smoothing:antialiased;cursor:pointer;text-rendering:auto}.br-disabled[_ngcontent-%COMP%]   .br-unit[_ngcontent-%COMP%], .br-readonly[_ngcontent-%COMP%]   .br-unit[_ngcontent-%COMP%]{cursor:default}"], changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BarRating, [{
        type: core.Component,
        args: [{
                selector: 'bar-rating',
                template: "<div class=\"br br-{{theme}}\"\n     [class.br-readonly]=\"readOnly\"\n     [class.br-disabled]=\"disabled\">\n\n  <div class=\"br-units\" (mouseleave)=\"reset()\">\n\n    <div class=\"br-unit\" *ngFor=\"let unit of contexts\"\n         [class.br-fraction]=\"unit.fraction\"\n         [class.br-selected]=\"unit.selected\"\n         [class.br-active]=\"unit.active\"\n         (click)=\"unit.click($event)\"\n         (mouseenter)=\"unit.enter()\"></div>\n  </div>\n\n  <div *ngIf=\"showText\" class=\"br-text\">{{ nextRate | rateTitle: titles }}</div>\n</div>\n",
                providers: [RATING_VALUE_ACCESSOR, RATING_VALUE_VALIDATOR],
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: ["*{box-sizing:border-box}.br{margin:15px 0;position:relative}.br-units{display:flex;flex-wrap:nowrap}.br-unit{-webkit-font-smoothing:antialiased;cursor:pointer;text-rendering:auto}.br-disabled .br-unit,.br-readonly .br-unit{cursor:default}"]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }]; }, { max: [{
            type: core.Input
        }], readOnly: [{
            type: core.Input
        }], theme: [{
            type: core.Input
        }], showText: [{
            type: core.Input
        }], titles: [{
            type: core.Input
        }], required: [{
            type: core.Input
        }], hover: [{
            type: core.Output
        }], leave: [{
            type: core.Output
        }], rateChange: [{
            type: core.Output
        }], rate: [{
            type: core.Input
        }] }); })();
        return BarRating;
    }());
    BarRating.ctorParameters = function () { return [
        { type: core.ChangeDetectorRef }
    ]; };
    BarRating.propDecorators = {
        rate: [{ type: core.Input }],
        max: [{ type: core.Input }],
        readOnly: [{ type: core.Input }],
        theme: [{ type: core.Input }],
        showText: [{ type: core.Input }],
        titles: [{ type: core.Input }],
        required: [{ type: core.Input }],
        hover: [{ type: core.Output }],
        leave: [{ type: core.Output }],
        rateChange: [{ type: core.Output }]
    };

    var BarRatingPipe = /** @class */ (function () {
        function BarRatingPipe() {
        }
        BarRatingPipe.prototype.transform = function (value, titles) {
            if (value === void 0) { value = 0; }
            /** Initialize value with 0 in case of undefined */
            return titles[value] || value + 1;
        };
BarRatingPipe.ɵfac = function BarRatingPipe_Factory(t) { return new (t || BarRatingPipe)(); };
BarRatingPipe.ɵpipe = ɵngcc0.ɵɵdefinePipe({ name: "rateTitle", type: BarRatingPipe, pure: true });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BarRatingPipe, [{
        type: core.Pipe,
        args: [{
                name: 'rateTitle'
            }]
    }], function () { return []; }, null); })();
        return BarRatingPipe;
    }());

    var BarRatingModule = /** @class */ (function () {
        function BarRatingModule() {
        }
BarRatingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: BarRatingModule });
BarRatingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function BarRatingModule_Factory(t) { return new (t || BarRatingModule)(); }, imports: [[
            common.CommonModule,
            forms.FormsModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(BarRatingModule, { declarations: [BarRating, BarRatingPipe], imports: [ɵngcc1.CommonModule, ɵngcc2.FormsModule], exports: [BarRating] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BarRatingModule, [{
        type: core.NgModule,
        args: [{
                declarations: [BarRating, BarRatingPipe],
                imports: [
                    common.CommonModule,
                    forms.FormsModule
                ],
                exports: [BarRating]
            }]
    }], function () { return []; }, null); })();
        return BarRatingModule;
    }());

    /**
     * Generated bundle index. Do not edit.
     */

    exports.BarRating = BarRating;
    exports.BarRatingModule = BarRatingModule;
    exports.ɵa = BarRatingPipe;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
