// ngx-chartist/esm5/lib/ngx-chartist.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxChartistComponent } from './ngx-chartist.component';
import { NgxChartistService } from './ngx-chartist.service';
import * as ɵngcc0 from '@angular/core';
var NgxChartistModule = /** @class */ (function () {
    function NgxChartistModule() {
    }
NgxChartistModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxChartistModule });
NgxChartistModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxChartistModule_Factory(t) { return new (t || NgxChartistModule)(); }, providers: [
        NgxChartistService
    ], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxChartistModule, { declarations: function () { return [NgxChartistComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxChartistComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxChartistModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [NgxChartistComponent],
                exports: [NgxChartistComponent],
                providers: [
                    NgxChartistService
                ]
            }]
    }], function () { return []; }, null); })();
    return NgxChartistModule;
}());
export { NgxChartistModule };
