// ngx-star-rating/esm5/lib/ngx-star-rating.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxStarRatingComponent } from './ngx-star-rating.component';
import * as ɵngcc0 from '@angular/core';
var NgxStarRatingModule = /** @class */ (function () {
    function NgxStarRatingModule() {
    }
NgxStarRatingModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxStarRatingModule });
NgxStarRatingModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxStarRatingModule_Factory(t) { return new (t || NgxStarRatingModule)(); }, imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxStarRatingModule, { declarations: function () { return [NgxStarRatingComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxStarRatingComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxStarRatingModule, [{
        type: NgModule,
        args: [{
                declarations: [NgxStarRatingComponent],
                imports: [
                    CommonModule
                ],
                exports: [NgxStarRatingComponent]
            }]
    }], function () { return []; }, null); })();
    return NgxStarRatingModule;
}());
export { NgxStarRatingModule };
