// ngx-star-rating/esm5/lib/ngx-star-rating.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function NgxStarRatingComponent_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    var _r4 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "input", 2);
    ɵngcc0.ɵɵlistener("click", function NgxStarRatingComponent_ng_template_1_Template_input_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r4); var star_r1 = ctx.$implicit; var ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.rate(star_r1); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelement(1, "label", 3);
} if (rf & 2) {
    var star_r1 = ctx.$implicit;
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassMapInterpolate1("star star-", star_r1, "");
    ɵngcc0.ɵɵpropertyInterpolate2("id", "star-", star_r1, "-", ctx_r0.id, "");
    ɵngcc0.ɵɵproperty("checked", ctx_r0.value == star_r1)("disabled", ctx_r0.disabled);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassMapInterpolate1("star star-", star_r1, "");
    ɵngcc0.ɵɵpropertyInterpolate2("for", "star-", star_r1, "-", ctx_r0.id, "");
} }
var NgxStarRatingComponent = /** @class */ (function () {
    function NgxStarRatingComponent() {
        this.stars = [5, 4, 3, 2, 1];
        this.propagateChange = (/**
         * @param {?} _
         * @return {?}
         */
        function (_) { });
        if (!this.disabled) {
            this.disabled = false;
        }
    }
    /**
     * @return {?}
     */
    NgxStarRatingComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} rate
     * @return {?}
     */
    NgxStarRatingComponent.prototype.rate = /**
     * @param {?} rate
     * @return {?}
     */
    function (rate) {
        if (!this.disabled) {
            this.propagateChange(rate);
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NgxStarRatingComponent.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    NgxStarRatingComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.propagateChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    NgxStarRatingComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) { };
    /** @nocollapse */
    NgxStarRatingComponent.ctorParameters = function () { return []; };
    NgxStarRatingComponent.propDecorators = {
        id: [{ type: Input }],
        disabled: [{ type: Input }]
    };
NgxStarRatingComponent.ɵfac = function NgxStarRatingComponent_Factory(t) { return new (t || NgxStarRatingComponent)(); };
NgxStarRatingComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgxStarRatingComponent, selectors: [["ngx-star-rating"]], inputs: { disabled: "disabled", id: "id" }, features: [ɵngcc0.ɵɵProvidersFeature([
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: forwardRef(( /**
                 * @return {?}
                 */function () { return NgxStarRatingComponent; })),
                multi: true
            }
        ])], decls: 2, vars: 1, consts: [["action", ""], ["ngFor", "", 3, "ngForOf"], ["type", "radio", "name", "star", 3, "id", "checked", "disabled", "click"], [3, "for"]], template: function NgxStarRatingComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "form", 0);
        ɵngcc0.ɵɵtemplate(1, NgxStarRatingComponent_ng_template_1_Template, 2, 12, "ng-template", 1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.stars);
    } }, directives: [ɵngcc1.NgForOf], styles: ["@charset \"UTF-8\";*[_ngcontent-%COMP%]{margin:0;padding:0;font-family:roboto}body[_ngcontent-%COMP%]{background:#000}.cont[_ngcontent-%COMP%]{width:93%;max-width:350px;text-align:center;margin:4% auto;padding:30px 0;background:#111;color:#eee;border-radius:5px;border:thin solid #444;overflow:hidden}hr[_ngcontent-%COMP%]{margin:20px;border:none;border-bottom:thin solid rgba(255,255,255,.1)}div.title[_ngcontent-%COMP%]{font-size:2em}h1[_ngcontent-%COMP%]   span[_ngcontent-%COMP%]{font-weight:300;color:#fd4}div.stars[_ngcontent-%COMP%]{width:270px;display:inline-block}input.star[_ngcontent-%COMP%]{display:none}label.star[_ngcontent-%COMP%]{float:right;padding:10px;font-size:36px;color:#444;transition:.2s}input.star[_ngcontent-%COMP%]:checked ~ label.star[_ngcontent-%COMP%]:before{content:\"\uF005\";color:#fd4;transition:.25s}input.star-5[_ngcontent-%COMP%]:checked ~ label.star[_ngcontent-%COMP%]:before{color:#fe7;text-shadow:0 0 20px #952}input.star-1[_ngcontent-%COMP%]:checked ~ label.star[_ngcontent-%COMP%]:before{color:#f62}label.star[_ngcontent-%COMP%]:hover{-webkit-transform:rotate(-15deg) scale(1.3);transform:rotate(-15deg) scale(1.3)}label.star[_ngcontent-%COMP%]:before{content:\"\uF006\";font-family:FontAwesome}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxStarRatingComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-star-rating',
                template: "<form action=\"\">\n  <ng-template ngFor let-star [ngForOf]=\"stars\" let-currentElementIndex=\"(index + 1)\">\n    <input class=\"star star-{{ star }}\" id=\"star-{{ star }}-{{ id }}\" type=\"radio\" name=\"star\" (click)=\"rate(star)\"\n      [checked]=\"value == star\" [disabled]=\"disabled\" />\n    <label class=\"star star-{{ star }}\" for=\"star-{{ star }}-{{ id }}\"></label>\n  </ng-template>\n</form>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(( /**
                         * @return {?}
                         */function () { return NgxStarRatingComponent; })),
                        multi: true
                    }
                ],
                styles: ["@charset \"UTF-8\";*{margin:0;padding:0;font-family:roboto}body{background:#000}.cont{width:93%;max-width:350px;text-align:center;margin:4% auto;padding:30px 0;background:#111;color:#eee;border-radius:5px;border:thin solid #444;overflow:hidden}hr{margin:20px;border:none;border-bottom:thin solid rgba(255,255,255,.1)}div.title{font-size:2em}h1 span{font-weight:300;color:#fd4}div.stars{width:270px;display:inline-block}input.star{display:none}label.star{float:right;padding:10px;font-size:36px;color:#444;transition:.2s}input.star:checked~label.star:before{content:\"\uF005\";color:#fd4;transition:.25s}input.star-5:checked~label.star:before{color:#fe7;text-shadow:0 0 20px #952}input.star-1:checked~label.star:before{color:#f62}label.star:hover{-webkit-transform:rotate(-15deg) scale(1.3);transform:rotate(-15deg) scale(1.3)}label.star:before{content:\"\uF006\";font-family:FontAwesome}"]
            }]
    }], function () { return []; }, { disabled: [{
            type: Input
        }], id: [{
            type: Input
        }] }); })();
    return NgxStarRatingComponent;
}());
export { NgxStarRatingComponent };
if (false) {
    /** @type {?} */
    NgxStarRatingComponent.prototype.onChange;
    /** @type {?} */
    NgxStarRatingComponent.prototype.value;
    /** @type {?} */
    NgxStarRatingComponent.prototype.stars;
    /** @type {?} */
    NgxStarRatingComponent.prototype.id;
    /** @type {?} */
    NgxStarRatingComponent.prototype.disabled;
    /**
     * @type {?}
     * @private
     */
    NgxStarRatingComponent.prototype.propagateChange;
}
