// ngx-chartist/fesm5/ngx-chartist.js
import { __spread, __read, __values } from 'tslib';
import { Injectable, NgModule, defineInjectable, Component, ElementRef, Input } from '@angular/core';
import * as Chartist from 'chartist';
import { precision, escapingMap, Pie, Bar, Line, Candle, FixedScaleAxis, AutoScaleAxis, StepAxis, Svg, Interpolation, noop, plugins, alphaNumerate, extend, replaceAll, ensureUnit, quantity, query, times, sum, mapMultiply, mapAdd, serialMap, roundWithPrecision, getMultiValue, serialize, deserialize, createSvg } from 'chartist';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as ɵngcc0 from '@angular/core';
var NgxChartistService = /** @class */ (function () {
    function NgxChartistService() {
    }
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getPrecision = /**
     * @return {?}
     */
    function () {
        return precision;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getEscapingMap = /**
     * @return {?}
     */
    function () {
        return escapingMap;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getPie = /**
     * @return {?}
     */
    function () {
        return Pie;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getBar = /**
     * @return {?}
     */
    function () {
        return Bar;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getLine = /**
     * @return {?}
     */
    function () {
        return Line;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getCandle = /**
     * @return {?}
     */
    function () {
        return Candle;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getFixedScaleAxis = /**
     * @return {?}
     */
    function () {
        return FixedScaleAxis;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getAutoScaleAxis = /**
     * @return {?}
     */
    function () {
        return AutoScaleAxis;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getStepAxis = /**
     * @return {?}
     */
    function () {
        return StepAxis;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getSvg = /**
     * @return {?}
     */
    function () {
        return Svg;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getInterpolation = /**
     * @return {?}
     */
    function () {
        return Interpolation;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getNoop = /**
     * @return {?}
     */
    function () {
        return noop;
    };
    /**
     * @return {?}
     */
    NgxChartistService.prototype.getPlugins = /**
     * @return {?}
     */
    function () {
        return plugins;
    };
    /**
     * @param {?} n
     * @return {?}
     */
    NgxChartistService.prototype.alphaNumerate = /**
     * @param {?} n
     * @return {?}
     */
    function (n) {
        return alphaNumerate(n);
    };
    /**
     * @param {?} target
     * @param {...?} sources
     * @return {?}
     */
    NgxChartistService.prototype.extend = /**
     * @param {?} target
     * @param {...?} sources
     * @return {?}
     */
    function (target) {
        var sources = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            sources[_i - 1] = arguments[_i];
        }
        return extend.apply(Chartist, __spread([target], sources));
    };
    /**
     * @param {?} str
     * @param {?} subStr
     * @param {?} newSubStr
     * @return {?}
     */
    NgxChartistService.prototype.replaceAll = /**
     * @param {?} str
     * @param {?} subStr
     * @param {?} newSubStr
     * @return {?}
     */
    function (str, subStr, newSubStr) {
        return replaceAll(str, subStr, newSubStr);
    };
    /**
     * @param {?} value
     * @param {?} unit
     * @return {?}
     */
    NgxChartistService.prototype.ensureUnit = /**
     * @param {?} value
     * @param {?} unit
     * @return {?}
     */
    function (value, unit) {
        return ensureUnit(value, unit);
    };
    /**
     * @param {?} input
     * @return {?}
     */
    NgxChartistService.prototype.quantity = /**
     * @param {?} input
     * @return {?}
     */
    function (input) {
        return quantity(input);
    };
    /**
     * @param {?} query
     * @return {?}
     */
    NgxChartistService.prototype.query = /**
     * @param {?} query
     * @return {?}
     */
    function (query$$1) {
        return query(query$$1);
    };
    /**
     * @param {?} length
     * @return {?}
     */
    NgxChartistService.prototype.times = /**
     * @param {?} length
     * @return {?}
     */
    function (length) {
        return times(length);
    };
    /**
     * @param {?} previous
     * @param {?} current
     * @return {?}
     */
    NgxChartistService.prototype.sum = /**
     * @param {?} previous
     * @param {?} current
     * @return {?}
     */
    function (previous, current) {
        return sum(previous, current);
    };
    /**
     * @param {?} factor
     * @return {?}
     */
    NgxChartistService.prototype.mapMultiply = /**
     * @param {?} factor
     * @return {?}
     */
    function (factor) {
        return mapMultiply(factor);
    };
    /**
     * @param {?} addend
     * @return {?}
     */
    NgxChartistService.prototype.mapAdd = /**
     * @param {?} addend
     * @return {?}
     */
    function (addend) {
        return mapAdd(addend);
    };
    /**
     * @param {?} arr
     * @param {?} cb
     * @return {?}
     */
    NgxChartistService.prototype.serialMap = /**
     * @param {?} arr
     * @param {?} cb
     * @return {?}
     */
    function (arr, cb) {
        return serialMap(arr, cb);
    };
    /**
     * @param {?} value
     * @param {?=} digits
     * @return {?}
     */
    NgxChartistService.prototype.roundWithPrecision = /**
     * @param {?} value
     * @param {?=} digits
     * @return {?}
     */
    function (value, digits) {
        return roundWithPrecision(value, digits);
    };
    /**
     * @param {?} value
     * @param {?=} dimension
     * @return {?}
     */
    NgxChartistService.prototype.getMultiValue = /**
     * @param {?} value
     * @param {?=} dimension
     * @return {?}
     */
    function (value, dimension) {
        return getMultiValue(value, dimension);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    NgxChartistService.prototype.serialize = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return serialize(data);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    NgxChartistService.prototype.deserialize = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return deserialize(data);
    };
    /**
     * @param {?} container
     * @param {?} width
     * @param {?} height
     * @param {?} className
     * @return {?}
     */
    NgxChartistService.prototype.createSvg = /**
     * @param {?} container
     * @param {?} width
     * @param {?} height
     * @param {?} className
     * @return {?}
     */
    function (container, width, height, className) {
        return createSvg(container, width, height, className);
    };
    NgxChartistService.ctorParameters = function () { return []; };
    /** @nocollapse */ NgxChartistService.ngInjectableDef = defineInjectable({ factory: function NgxChartistService_Factory() { return new NgxChartistService(); }, token: NgxChartistService, providedIn: "root" });
NgxChartistService.ɵfac = function NgxChartistService_Factory(t) { return new (t || NgxChartistService)(); };
NgxChartistService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: NgxChartistService, factory: function (t) { return NgxChartistService.ɵfac(t); }, providedIn: 'root' });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxChartistService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return []; }, null); })();
    return NgxChartistService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var NgxChartistComponent = /** @class */ (function () {
    function NgxChartistComponent(element) {
        this.element = element.nativeElement;
    }
    /**
     * @return {?}
     */
    NgxChartistComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.type || !this.data) {
            Promise.reject('Expected at least type and data.');
        }
        return this.renderChart().then(function (chart) {
            if (_this.events !== undefined) {
                _this.bindEvents(chart);
            }
            return chart;
        });
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    NgxChartistComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (this.element) {
            this.update(changes);
        }
    };
    /**
     * @return {?}
     */
    NgxChartistComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.chart) {
            this.chart.detach();
        }
    };
    /**
     * @return {?}
     */
    NgxChartistComponent.prototype.renderChart = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var promises = [
            this.type,
            this.element,
            this.data,
            this.options,
            this.responsiveOptions
        ];
        if (this.lineSmooth) {
            switch (this.lineSmooth['interpolation']) {
                case 'cardinal': {
                    this.options['lineSmooth'] = Interpolation.cardinal(this.lineSmooth['data']);
                    break;
                }
                case 'simple': {
                    this.options['lineSmooth'] = Interpolation.simple(this.lineSmooth['data']);
                    break;
                }
                case 'none': {
                    this.options['lineSmooth'] = Interpolation.none(this.lineSmooth['data']);
                    break;
                }
                case 'step': {
                    this.options['lineSmooth'] = Interpolation.step(this.lineSmooth['data']);
                    break;
                }
            }
        }
        return Promise.all(promises).then(function (values) {
            var _a = __read(values), type = _a[0], args = _a.slice(1);
            if (!(type in Chartist)) {
                throw new Error(type + " is not a valid chart type");
            }
            _this.chart = ((/** @type {?} */ (Chartist)))[type].apply(((/** @type {?} */ (Chartist))), __spread(args));
            return _this.chart;
        });
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    NgxChartistComponent.prototype.update = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!this.chart || 'type' in changes) {
            this.renderChart();
        }
        else {
            if (changes.data) {
                this.data = changes.data.currentValue;
            }
            if (changes.options) {
                this.options = changes.options.currentValue;
            }
            ((/** @type {?} */ (this.chart))).update(this.data, this.options);
        }
    };
    /**
     * @param {?} chart
     * @return {?}
     */
    NgxChartistComponent.prototype.bindEvents = /**
     * @param {?} chart
     * @return {?}
     */
    function (chart) {
        var e_1, _a;
        try {
            for (var _b = __values(Object.keys(this.events)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var event_1 = _c.value;
                chart.on(event_1, this.events[event_1]);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    NgxChartistComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    NgxChartistComponent.propDecorators = {
        data: [{ type: Input }],
        type: [{ type: Input }],
        options: [{ type: Input }],
        lineSmooth: [{ type: Input }],
        responsiveOptions: [{ type: Input }],
        events: [{ type: Input }]
    };
NgxChartistComponent.ɵfac = function NgxChartistComponent_Factory(t) { return new (t || NgxChartistComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
NgxChartistComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgxChartistComponent, selectors: [["lib-ngx-chartist"]], inputs: { data: "data", options: "options", type: "type", lineSmooth: "lineSmooth", responsiveOptions: "responsiveOptions", events: "events" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 0, vars: 0, template: function NgxChartistComponent_Template(rf, ctx) { }, styles: ["[_nghost-%COMP%]:not(.without-styles)     .ct-label{fill:rgba(0,0,0,.4);color:rgba(0,0,0,.4);font-size:.75rem;line-height:1}[_nghost-%COMP%]:not(.without-styles)     .ct-chart-bar .ct-label, [_nghost-%COMP%]:not(.without-styles)     .ct-chart-line .ct-label{display:block;display:flex}[_nghost-%COMP%]:not(.without-styles)     .ct-chart-donut .ct-label, [_nghost-%COMP%]:not(.without-styles)     .ct-chart-pie .ct-label{dominant-baseline:central}[_nghost-%COMP%]:not(.without-styles)     .ct-label.ct-horizontal.ct-start{align-items:flex-end;justify-content:flex-start;text-align:left;text-anchor:start}[_nghost-%COMP%]:not(.without-styles)     .ct-label.ct-horizontal.ct-end{align-items:flex-start;justify-content:flex-start;text-align:left;text-anchor:start}[_nghost-%COMP%]:not(.without-styles)     .ct-label.ct-vertical.ct-start{align-items:flex-end;justify-content:flex-end;text-align:right;text-anchor:end}[_nghost-%COMP%]:not(.without-styles)     .ct-label.ct-vertical.ct-end{align-items:flex-end;justify-content:flex-start;text-align:left;text-anchor:start}[_nghost-%COMP%]:not(.without-styles)     .ct-chart-bar .ct-label.ct-horizontal.ct-start{align-items:flex-end;justify-content:center;text-align:center;text-anchor:start}[_nghost-%COMP%]:not(.without-styles)     .ct-chart-bar .ct-label.ct-horizontal.ct-end{align-items:flex-start;justify-content:center;text-align:center;text-anchor:start}[_nghost-%COMP%]:not(.without-styles)     .ct-chart-bar.ct-horizontal-bars .ct-label.ct-horizontal.ct-start{align-items:flex-end;justify-content:flex-start;text-align:left;text-anchor:start}[_nghost-%COMP%]:not(.without-styles)     .ct-chart-bar.ct-horizontal-bars .ct-label.ct-horizontal.ct-end{align-items:flex-start;justify-content:flex-start;text-align:left;text-anchor:start}[_nghost-%COMP%]:not(.without-styles)     .ct-chart-bar.ct-horizontal-bars .ct-label.ct-vertical.ct-start{align-items:center;justify-content:flex-end;text-align:right;text-anchor:end}[_nghost-%COMP%]:not(.without-styles)     .ct-chart-bar.ct-horizontal-bars .ct-label.ct-vertical.ct-end{align-items:center;justify-content:flex-start;text-align:left;text-anchor:end}[_nghost-%COMP%]:not(.without-styles)     .ct-grid{stroke:rgba(0,0,0,.2);stroke-width:1px;stroke-dasharray:2px}[_nghost-%COMP%]:not(.without-styles)     .ct-grid-background{fill:none}[_nghost-%COMP%]:not(.without-styles)     .ct-point{stroke-width:10px;stroke-linecap:round}[_nghost-%COMP%]:not(.without-styles)     .ct-line{fill:none;stroke-width:4px}[_nghost-%COMP%]:not(.without-styles)     .ct-area{stroke:none;fill-opacity:.1}[_nghost-%COMP%]:not(.without-styles)     .ct-bar{fill:none;stroke-width:10px}[_nghost-%COMP%]:not(.without-styles)     .ct-slice-donut{fill:none;stroke-width:60px}[_nghost-%COMP%]:not(.without-styles)     .ct-series-a .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-a .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-a .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-a .ct-slice-donut{stroke:#d70206}[_nghost-%COMP%]:not(.without-styles)     .ct-series-a .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-a .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-a .ct-slice-pie{fill:#d70206}[_nghost-%COMP%]:not(.without-styles)     .ct-series-b .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-b .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-b .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-b .ct-slice-donut{stroke:#f05b4f}[_nghost-%COMP%]:not(.without-styles)     .ct-series-b .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-b .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-b .ct-slice-pie{fill:#f05b4f}[_nghost-%COMP%]:not(.without-styles)     .ct-series-c .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-c .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-c .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-c .ct-slice-donut{stroke:#f4c63d}[_nghost-%COMP%]:not(.without-styles)     .ct-series-c .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-c .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-c .ct-slice-pie{fill:#f4c63d}[_nghost-%COMP%]:not(.without-styles)     .ct-series-d .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-d .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-d .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-d .ct-slice-donut{stroke:#d17905}[_nghost-%COMP%]:not(.without-styles)     .ct-series-d .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-d .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-d .ct-slice-pie{fill:#d17905}[_nghost-%COMP%]:not(.without-styles)     .ct-series-e .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-e .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-e .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-e .ct-slice-donut{stroke:#453d3f}[_nghost-%COMP%]:not(.without-styles)     .ct-series-e .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-e .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-e .ct-slice-pie{fill:#453d3f}[_nghost-%COMP%]:not(.without-styles)     .ct-series-f .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-f .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-f .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-f .ct-slice-donut{stroke:#59922b}[_nghost-%COMP%]:not(.without-styles)     .ct-series-f .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-f .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-f .ct-slice-pie{fill:#59922b}[_nghost-%COMP%]:not(.without-styles)     .ct-series-g .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-g .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-g .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-g .ct-slice-donut{stroke:#0544d3}[_nghost-%COMP%]:not(.without-styles)     .ct-series-g .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-g .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-g .ct-slice-pie{fill:#0544d3}[_nghost-%COMP%]:not(.without-styles)     .ct-series-h .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-h .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-h .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-h .ct-slice-donut{stroke:#6b0392}[_nghost-%COMP%]:not(.without-styles)     .ct-series-h .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-h .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-h .ct-slice-pie{fill:#6b0392}[_nghost-%COMP%]:not(.without-styles)     .ct-series-i .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-i .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-i .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-i .ct-slice-donut{stroke:#f05b4f}[_nghost-%COMP%]:not(.without-styles)     .ct-series-i .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-i .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-i .ct-slice-pie{fill:#f05b4f}[_nghost-%COMP%]:not(.without-styles)     .ct-series-j .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-j .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-j .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-j .ct-slice-donut{stroke:#dda458}[_nghost-%COMP%]:not(.without-styles)     .ct-series-j .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-j .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-j .ct-slice-pie{fill:#dda458}[_nghost-%COMP%]:not(.without-styles)     .ct-series-k .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-k .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-k .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-k .ct-slice-donut{stroke:#eacf7d}[_nghost-%COMP%]:not(.without-styles)     .ct-series-k .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-k .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-k .ct-slice-pie{fill:#eacf7d}[_nghost-%COMP%]:not(.without-styles)     .ct-series-l .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-l .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-l .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-l .ct-slice-donut{stroke:#86797d}[_nghost-%COMP%]:not(.without-styles)     .ct-series-l .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-l .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-l .ct-slice-pie{fill:#86797d}[_nghost-%COMP%]:not(.without-styles)     .ct-series-m .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-m .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-m .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-m .ct-slice-donut{stroke:#b2c326}[_nghost-%COMP%]:not(.without-styles)     .ct-series-m .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-m .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-m .ct-slice-pie{fill:#b2c326}[_nghost-%COMP%]:not(.without-styles)     .ct-series-n .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-n .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-n .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-n .ct-slice-donut{stroke:#6188e2}[_nghost-%COMP%]:not(.without-styles)     .ct-series-n .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-n .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-n .ct-slice-pie{fill:#6188e2}[_nghost-%COMP%]:not(.without-styles)     .ct-series-o .ct-bar, [_nghost-%COMP%]:not(.without-styles)     .ct-series-o .ct-line, [_nghost-%COMP%]:not(.without-styles)     .ct-series-o .ct-point, [_nghost-%COMP%]:not(.without-styles)     .ct-series-o .ct-slice-donut{stroke:#a748ca}[_nghost-%COMP%]:not(.without-styles)     .ct-series-o .ct-area, [_nghost-%COMP%]:not(.without-styles)     .ct-series-o .ct-slice-donut-solid, [_nghost-%COMP%]:not(.without-styles)     .ct-series-o .ct-slice-pie{fill:#a748ca}[_nghost-%COMP%]:not(.without-styles)     .ct-square{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-square:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-square:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-square>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-second{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-second:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:93.75%}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-second:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-second>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-major-second{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-second:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:88.88889%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-second:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-major-second>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-third{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-third:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:83.33333%}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-third:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-third>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-major-third{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-third:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:80%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-third:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-major-third>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-perfect-fourth{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-perfect-fourth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:75%}[_nghost-%COMP%]:not(.without-styles)     .ct-perfect-fourth:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-perfect-fourth>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-perfect-fifth{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-perfect-fifth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:66.66667%}[_nghost-%COMP%]:not(.without-styles)     .ct-perfect-fifth:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-perfect-fifth>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-sixth{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-sixth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:62.5%}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-sixth:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-sixth>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-golden-section{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-golden-section:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:61.8047%}[_nghost-%COMP%]:not(.without-styles)     .ct-golden-section:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-golden-section>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-major-sixth{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-sixth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:60%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-sixth:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-major-sixth>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-seventh{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-seventh:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:56.25%}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-seventh:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-minor-seventh>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-major-seventh{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-seventh:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:53.33333%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-seventh:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-major-seventh>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-octave{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-octave:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:50%}[_nghost-%COMP%]:not(.without-styles)     .ct-octave:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-octave>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-major-tenth{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-tenth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:40%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-tenth:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-major-tenth>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-major-eleventh{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-eleventh:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:37.5%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-eleventh:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-major-eleventh>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-major-twelfth{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-twelfth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:33.33333%}[_nghost-%COMP%]:not(.without-styles)     .ct-major-twelfth:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-major-twelfth>svg{display:block;position:absolute;top:0;left:0}[_nghost-%COMP%]:not(.without-styles)     .ct-double-octave{display:block;position:relative;width:100%}[_nghost-%COMP%]:not(.without-styles)     .ct-double-octave:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:25%}[_nghost-%COMP%]:not(.without-styles)     .ct-double-octave:after{content:\"\";display:table;clear:both}[_nghost-%COMP%]:not(.without-styles)     .ct-double-octave>svg{display:block;position:absolute;top:0;left:0}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxChartistComponent, [{
        type: Component,
        args: [{
                selector: 'lib-ngx-chartist',
                template: '',
                styles: [":host:not(.without-styles) ::ng-deep .ct-label{fill:rgba(0,0,0,.4);color:rgba(0,0,0,.4);font-size:.75rem;line-height:1}:host:not(.without-styles) ::ng-deep .ct-chart-bar .ct-label,:host:not(.without-styles) ::ng-deep .ct-chart-line .ct-label{display:block;display:flex}:host:not(.without-styles) ::ng-deep .ct-chart-donut .ct-label,:host:not(.without-styles) ::ng-deep .ct-chart-pie .ct-label{dominant-baseline:central}:host:not(.without-styles) ::ng-deep .ct-label.ct-horizontal.ct-start{align-items:flex-end;justify-content:flex-start;text-align:left;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-label.ct-horizontal.ct-end{align-items:flex-start;justify-content:flex-start;text-align:left;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-label.ct-vertical.ct-start{align-items:flex-end;justify-content:flex-end;text-align:right;text-anchor:end}:host:not(.without-styles) ::ng-deep .ct-label.ct-vertical.ct-end{align-items:flex-end;justify-content:flex-start;text-align:left;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-chart-bar .ct-label.ct-horizontal.ct-start{align-items:flex-end;justify-content:center;text-align:center;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-chart-bar .ct-label.ct-horizontal.ct-end{align-items:flex-start;justify-content:center;text-align:center;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-chart-bar.ct-horizontal-bars .ct-label.ct-horizontal.ct-start{align-items:flex-end;justify-content:flex-start;text-align:left;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-chart-bar.ct-horizontal-bars .ct-label.ct-horizontal.ct-end{align-items:flex-start;justify-content:flex-start;text-align:left;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-chart-bar.ct-horizontal-bars .ct-label.ct-vertical.ct-start{align-items:center;justify-content:flex-end;text-align:right;text-anchor:end}:host:not(.without-styles) ::ng-deep .ct-chart-bar.ct-horizontal-bars .ct-label.ct-vertical.ct-end{align-items:center;justify-content:flex-start;text-align:left;text-anchor:end}:host:not(.without-styles) ::ng-deep .ct-grid{stroke:rgba(0,0,0,.2);stroke-width:1px;stroke-dasharray:2px}:host:not(.without-styles) ::ng-deep .ct-grid-background{fill:none}:host:not(.without-styles) ::ng-deep .ct-point{stroke-width:10px;stroke-linecap:round}:host:not(.without-styles) ::ng-deep .ct-line{fill:none;stroke-width:4px}:host:not(.without-styles) ::ng-deep .ct-area{stroke:none;fill-opacity:.1}:host:not(.without-styles) ::ng-deep .ct-bar{fill:none;stroke-width:10px}:host:not(.without-styles) ::ng-deep .ct-slice-donut{fill:none;stroke-width:60px}:host:not(.without-styles) ::ng-deep .ct-series-a .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-a .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-a .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-a .ct-slice-donut{stroke:#d70206}:host:not(.without-styles) ::ng-deep .ct-series-a .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-a .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-a .ct-slice-pie{fill:#d70206}:host:not(.without-styles) ::ng-deep .ct-series-b .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-b .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-b .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-b .ct-slice-donut{stroke:#f05b4f}:host:not(.without-styles) ::ng-deep .ct-series-b .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-b .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-b .ct-slice-pie{fill:#f05b4f}:host:not(.without-styles) ::ng-deep .ct-series-c .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-c .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-c .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-c .ct-slice-donut{stroke:#f4c63d}:host:not(.without-styles) ::ng-deep .ct-series-c .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-c .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-c .ct-slice-pie{fill:#f4c63d}:host:not(.without-styles) ::ng-deep .ct-series-d .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-d .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-d .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-d .ct-slice-donut{stroke:#d17905}:host:not(.without-styles) ::ng-deep .ct-series-d .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-d .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-d .ct-slice-pie{fill:#d17905}:host:not(.without-styles) ::ng-deep .ct-series-e .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-e .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-e .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-e .ct-slice-donut{stroke:#453d3f}:host:not(.without-styles) ::ng-deep .ct-series-e .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-e .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-e .ct-slice-pie{fill:#453d3f}:host:not(.without-styles) ::ng-deep .ct-series-f .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-f .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-f .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-f .ct-slice-donut{stroke:#59922b}:host:not(.without-styles) ::ng-deep .ct-series-f .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-f .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-f .ct-slice-pie{fill:#59922b}:host:not(.without-styles) ::ng-deep .ct-series-g .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-g .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-g .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-g .ct-slice-donut{stroke:#0544d3}:host:not(.without-styles) ::ng-deep .ct-series-g .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-g .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-g .ct-slice-pie{fill:#0544d3}:host:not(.without-styles) ::ng-deep .ct-series-h .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-h .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-h .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-h .ct-slice-donut{stroke:#6b0392}:host:not(.without-styles) ::ng-deep .ct-series-h .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-h .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-h .ct-slice-pie{fill:#6b0392}:host:not(.without-styles) ::ng-deep .ct-series-i .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-i .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-i .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-i .ct-slice-donut{stroke:#f05b4f}:host:not(.without-styles) ::ng-deep .ct-series-i .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-i .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-i .ct-slice-pie{fill:#f05b4f}:host:not(.without-styles) ::ng-deep .ct-series-j .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-j .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-j .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-j .ct-slice-donut{stroke:#dda458}:host:not(.without-styles) ::ng-deep .ct-series-j .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-j .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-j .ct-slice-pie{fill:#dda458}:host:not(.without-styles) ::ng-deep .ct-series-k .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-k .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-k .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-k .ct-slice-donut{stroke:#eacf7d}:host:not(.without-styles) ::ng-deep .ct-series-k .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-k .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-k .ct-slice-pie{fill:#eacf7d}:host:not(.without-styles) ::ng-deep .ct-series-l .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-l .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-l .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-l .ct-slice-donut{stroke:#86797d}:host:not(.without-styles) ::ng-deep .ct-series-l .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-l .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-l .ct-slice-pie{fill:#86797d}:host:not(.without-styles) ::ng-deep .ct-series-m .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-m .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-m .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-m .ct-slice-donut{stroke:#b2c326}:host:not(.without-styles) ::ng-deep .ct-series-m .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-m .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-m .ct-slice-pie{fill:#b2c326}:host:not(.without-styles) ::ng-deep .ct-series-n .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-n .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-n .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-n .ct-slice-donut{stroke:#6188e2}:host:not(.without-styles) ::ng-deep .ct-series-n .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-n .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-n .ct-slice-pie{fill:#6188e2}:host:not(.without-styles) ::ng-deep .ct-series-o .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-o .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-o .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-o .ct-slice-donut{stroke:#a748ca}:host:not(.without-styles) ::ng-deep .ct-series-o .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-o .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-o .ct-slice-pie{fill:#a748ca}:host:not(.without-styles) ::ng-deep .ct-square{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-square:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:100%}:host:not(.without-styles) ::ng-deep .ct-square:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-square>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-minor-second{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-minor-second:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:93.75%}:host:not(.without-styles) ::ng-deep .ct-minor-second:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-minor-second>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-second{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-second:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:88.88889%}:host:not(.without-styles) ::ng-deep .ct-major-second:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-second>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-minor-third{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-minor-third:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:83.33333%}:host:not(.without-styles) ::ng-deep .ct-minor-third:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-minor-third>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-third{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-third:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:80%}:host:not(.without-styles) ::ng-deep .ct-major-third:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-third>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-perfect-fourth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-perfect-fourth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:75%}:host:not(.without-styles) ::ng-deep .ct-perfect-fourth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-perfect-fourth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-perfect-fifth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-perfect-fifth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:66.66667%}:host:not(.without-styles) ::ng-deep .ct-perfect-fifth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-perfect-fifth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-minor-sixth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-minor-sixth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:62.5%}:host:not(.without-styles) ::ng-deep .ct-minor-sixth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-minor-sixth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-golden-section{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-golden-section:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:61.8047%}:host:not(.without-styles) ::ng-deep .ct-golden-section:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-golden-section>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-sixth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-sixth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:60%}:host:not(.without-styles) ::ng-deep .ct-major-sixth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-sixth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-minor-seventh{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-minor-seventh:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:56.25%}:host:not(.without-styles) ::ng-deep .ct-minor-seventh:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-minor-seventh>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-seventh{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-seventh:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:53.33333%}:host:not(.without-styles) ::ng-deep .ct-major-seventh:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-seventh>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-octave{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-octave:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:50%}:host:not(.without-styles) ::ng-deep .ct-octave:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-octave>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-tenth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-tenth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:40%}:host:not(.without-styles) ::ng-deep .ct-major-tenth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-tenth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-eleventh{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-eleventh:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:37.5%}:host:not(.without-styles) ::ng-deep .ct-major-eleventh:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-eleventh>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-twelfth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-twelfth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:33.33333%}:host:not(.without-styles) ::ng-deep .ct-major-twelfth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-twelfth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-double-octave{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-double-octave:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:25%}:host:not(.without-styles) ::ng-deep .ct-double-octave:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-double-octave>svg{display:block;position:absolute;top:0;left:0}"]
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { data: [{
            type: Input
        }], options: [{
            type: Input
        }], type: [{
            type: Input
        }], lineSmooth: [{
            type: Input
        }], responsiveOptions: [{
            type: Input
        }], events: [{
            type: Input
        }] }); })();
    return NgxChartistComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var NgxChartistModule = /** @class */ (function () {
    function NgxChartistModule() {
    }
NgxChartistModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxChartistModule });
NgxChartistModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxChartistModule_Factory(t) { return new (t || NgxChartistModule)(); }, providers: [
        NgxChartistService
    ], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxChartistModule, { declarations: function () { return [NgxChartistComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxChartistComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxChartistModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule
                ],
                declarations: [NgxChartistComponent],
                exports: [NgxChartistComponent],
                providers: [
                    NgxChartistService
                ]
            }]
    }], function () { return []; }, null); })();
    return NgxChartistModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { NgxChartistService, NgxChartistComponent, NgxChartistModule };
