// ngx-notification-msg/esm5/lib/ngx-notification-msg.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function NgxNotificationMsgComponent_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 15);
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(1, "svg");
    ɵngcc0.ɵɵelement(2, "use");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵattribute("href", ctx_r0.statusToClass[ctx_r0.status], null, "xlink");
} }
function NgxNotificationMsgComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵtext(1);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r1.msgs, " ");
} }
function NgxNotificationMsgComponent_div_6_ul_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul");
    ɵngcc0.ɵɵelementStart(1, "li");
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var msg_r5 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(msg_r5);
} }
function NgxNotificationMsgComponent_div_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 16);
    ɵngcc0.ɵɵtemplate(1, NgxNotificationMsgComponent_div_6_ul_1_Template, 3, 1, "ul", 17);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ctx_r2.msgs);
} }
function NgxNotificationMsgComponent_button_7_Template(rf, ctx) { if (rf & 1) {
    var _r7 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 18);
    ɵngcc0.ɵɵlistener("click", function NgxNotificationMsgComponent_button_7_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r7); var ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.close(); });
    ɵngcc0.ɵɵnamespaceSVG();
    ɵngcc0.ɵɵelementStart(1, "svg");
    ɵngcc0.ɵɵelement(2, "use");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵattribute("href", "#clear", null, "xlink");
} }
var NgxNotificationMsgComponent = /** @class */ (function () {
    function NgxNotificationMsgComponent(cd, element) {
        var _a;
        this.cd = cd;
        this.element = element;
        this.status = NgxNotificationStatusMsg.NONE;
        this.direction = NgxNotificationDirection.TOP;
        this.displayProgressBar = true;
        this.displayIcon = true;
        this.delay = 3000;
        this.closeable = true;
        this.destroy = new EventEmitter();
        this.componentState = NgxNotificationMsgComponentState.CLOSE;
        this.componentStates = NgxNotificationMsgComponentState;
        this.statusToClass = (_a = {},
            _a[NgxNotificationStatusMsg.NONE] = '',
            _a[NgxNotificationStatusMsg.INFO] = '#0067FF',
            _a[NgxNotificationStatusMsg.FAILURE] = '#FE355A',
            _a[NgxNotificationStatusMsg.SUCCESS] = '#00CC69',
            _a);
        this.msgs = [];
        this.NONE = 'NONE';
    }
    Object.defineProperty(NgxNotificationMsgComponent.prototype, "messages", {
        set: /**
         * @param {?} messages
         * @return {?}
         */
        function (messages) {
            this.msgs = tslib_1.__spread(messages);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.init();
    };
    /**
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.componentState = NgxNotificationMsgComponentState.OPEN;
            _this.cd.markForCheck();
        }));
    };
    /**
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.isNotificationMsgOpened = /**
     * @return {?}
     */
    function () {
        return this.componentState === this.componentStates.OPEN;
    };
    /**
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.isProgressBarDisplayed = /**
     * @return {?}
     */
    function () {
        return this.status !== this.NONE && this.displayProgressBar;
    };
    /**
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.isIconDisplayed = /**
     * @return {?}
     */
    function () {
        return this.status !== this.NONE && this.displayIcon;
    };
    /**
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.getPosition = /**
     * @return {?}
     */
    function () {
        return tslib_1.__assign({}, this.getDefaultPosition(), (this.componentState === this.componentStates.OPEN && this.getDynamicPosition()));
    };
    /**
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.mouseEnter = /**
     * @return {?}
     */
    function () {
        this.mouseEnterTimestamp = performance.now();
        clearTimeout(this.closeTimeout);
        clearTimeout(this.destroyTimeout);
    };
    /**
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.mouseLeave = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var timestampGap = this.mouseEnterTimestamp - this.referencePointTimestamp;
        this.autoSelfDestroy(this.delay - timestampGap);
        this.referencePointTimestamp = performance.now() - timestampGap;
    };
    /**
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.componentState = NgxNotificationMsgComponentState.CLOSE;
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.destroy.emit();
        }), NgxNotificationMsgComponent.DELAY_ON_CLICK);
    };
    /**
     * @private
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.init = /**
     * @private
     * @return {?}
     */
    function () {
        this.referencePointTimestamp = performance.now();
        this.initTheme();
        this.autoSelfDestroy(this.delay);
    };
    /**
     * @private
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.initTheme = /**
     * @private
     * @return {?}
     */
    function () {
        this.element.nativeElement.style.setProperty('--ngx-notification-msg-delay', this.delay + "ms");
        this.element.nativeElement.style.setProperty('--ngx-notification-msg-color', this.statusToClass[this.status]);
    };
    /**
     * @private
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.getDefaultPosition = /**
     * @private
     * @return {?}
     */
    function () {
        switch (this.direction) {
            case NgxNotificationDirection.TOP:
                return { top: '0', right: '50%', transform: "translate(50%, -100%)" };
            case NgxNotificationDirection.TOP_RIGHT:
                return { top: '0', right: '20px', transform: "translateY(-100%)" };
            case NgxNotificationDirection.TOP_LEFT:
                return { top: '0', left: '20px', transform: "translateY(-100%)" };
            case NgxNotificationDirection.BOTTOM:
                return { bottom: '0', right: '50%', transform: "translateX(50%)" };
            case NgxNotificationDirection.BOTTOM_RIGHT:
                return { bottom: '0', right: '20px', transform: "translateY(100%)" };
            case NgxNotificationDirection.BOTTOM_LEFT:
                return { bottom: '0', left: '20px', transform: "translateY(100%)" };
        }
    };
    /**
     * @private
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.getDynamicPosition = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var top = "calc(" + 100 * this.index + "% + " + 20 * (this.index + 1) + "px)";
        /** @type {?} */
        var bottom = "calc(" + -100 * (this.index) + "% + " + -20 * (this.index + 1) + "px)";
        switch (this.direction) {
            case NgxNotificationDirection.TOP:
                return { transform: "translate(50%, " + top + ")" };
            case NgxNotificationDirection.TOP_RIGHT:
                return { transform: "translateY(" + top + ")" };
            case NgxNotificationDirection.TOP_LEFT:
                return { transform: "translateY(" + top + ")" };
            case NgxNotificationDirection.BOTTOM:
                return { transform: "translate(50%, " + bottom + ")" };
            case NgxNotificationDirection.BOTTOM_RIGHT:
                return { transform: "translateY(" + bottom + ")" };
            case NgxNotificationDirection.BOTTOM_LEFT:
                return { transform: "translateY(" + bottom + ")" };
        }
    };
    /**
     * @private
     * @param {?} delay
     * @return {?}
     */
    NgxNotificationMsgComponent.prototype.autoSelfDestroy = /**
     * @private
     * @param {?} delay
     * @return {?}
     */
    function (delay) {
        var _this = this;
        this.closeTimeout = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.componentState = NgxNotificationMsgComponentState.CLOSE;
            _this.cd.markForCheck();
        }), delay);
        this.destroyTimeout = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.destroy.emit();
        }), delay + NgxNotificationMsgComponent.DELAY_ON_CLICK);
    };
    NgxNotificationMsgComponent.DELAY_ON_CLICK = 400;
    /** @nocollapse */
    NgxNotificationMsgComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: ElementRef }
    ]; };
    NgxNotificationMsgComponent.propDecorators = {
        status: [{ type: Input }],
        direction: [{ type: Input }],
        displayProgressBar: [{ type: Input }],
        displayIcon: [{ type: Input }],
        header: [{ type: Input }],
        messages: [{ type: Input }],
        delay: [{ type: Input }],
        closeable: [{ type: Input }],
        index: [{ type: Input }],
        destroy: [{ type: Output }]
    };
NgxNotificationMsgComponent.ɵfac = function NgxNotificationMsgComponent_Factory(t) { return new (t || NgxNotificationMsgComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
NgxNotificationMsgComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgxNotificationMsgComponent, selectors: [["ngx-notification-msg"]], inputs: { status: "status", direction: "direction", displayProgressBar: "displayProgressBar", displayIcon: "displayIcon", delay: "delay", closeable: "closeable", messages: "messages", header: "header", index: "index" }, outputs: { destroy: "destroy" }, decls: 20, vars: 10, consts: [[1, "ngx_notification-msg", 3, "ngStyle", "mouseenter", "mouseleave"], ["class", "ngx_notification-msg-icon-wrapper", 4, "ngIf"], [1, "ngx_notification-msg-content"], [1, "ngx_notification-msg-header"], ["class", "ngx_notification-msg-body", 4, "ngIf"], ["class", "ngx_notification-msg-close-button", 3, "click", 4, "ngIf"], ["display", "none"], ["width", "24", "height", "24", "viewBox", "0 0 32 32", "id", "00CC69"], ["d", "M16,0 C24.836556,0 32,7.163444 32,16 C32,24.836556 24.836556,32 16,32 C7.163444,32 0,24.836556 0,16 C0,7.163444 7.163444,0 16,0 Z M23.6,9.6 L13,20.2 L8.8,16 L7.4,17.4 L13,23 L25,11 L23.6,9.6 Z"], ["width", "24", "height", "24", "viewBox", "0 0 32 32", "id", "FE355A"], ["d", "M16,0 C24.836556,0 32,7.163444 32,16 C32,24.836556 24.836556,32 16,32 C7.163444,32 0,24.836556 0,16 C0,7.163444 7.163444,0 16,0 Z M21.59,9 L16,14.59 L10.41,9 L9,10.41 L14.59,16 L9,21.59 L10.41,23 L16,17.41 L21.59,23 L23,21.59 L17.41,16 L23,10.41 L21.59,9 Z"], ["width", "24", "height", "24", "viewBox", "0 0 32 32", "id", "0067FF"], ["d", "M16,0 C24.836556,0 32,7.163444 32,16 C32,24.836556 24.836556,32 16,32 C7.163444,32 0,24.836556 0,16 C0,7.163444 7.163444,0 16,0 Z M17,12 L15,12 L15,24 L17,24 L17,12 Z M17,8 L15,8 L15,10 L17,10 L17,8 Z"], ["viewBox", "0 0 24 24", "id", "clear"], ["d", "M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"], [1, "ngx_notification-msg-icon-wrapper"], [1, "ngx_notification-msg-body"], [4, "ngFor", "ngForOf"], [1, "ngx_notification-msg-close-button", 3, "click"]], template: function NgxNotificationMsgComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵlistener("mouseenter", function NgxNotificationMsgComponent_Template_div_mouseenter_0_listener() { return ctx.mouseEnter(); })("mouseleave", function NgxNotificationMsgComponent_Template_div_mouseleave_0_listener() { return ctx.mouseLeave(); });
        ɵngcc0.ɵɵtemplate(1, NgxNotificationMsgComponent_div_1_Template, 3, 1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelementStart(3, "div", 3);
        ɵngcc0.ɵɵtext(4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, NgxNotificationMsgComponent_div_5_Template, 2, 1, "div", 4);
        ɵngcc0.ɵɵtemplate(6, NgxNotificationMsgComponent_div_6_Template, 2, 1, "div", 4);
        ɵngcc0.ɵɵtemplate(7, NgxNotificationMsgComponent_button_7_Template, 3, 1, "button", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵnamespaceSVG();
        ɵngcc0.ɵɵelementStart(8, "svg", 6);
        ɵngcc0.ɵɵelementStart(9, "symbol", 7);
        ɵngcc0.ɵɵelement(10, "path", 8);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(11, "svg", 6);
        ɵngcc0.ɵɵelementStart(12, "symbol", 9);
        ɵngcc0.ɵɵelement(13, "path", 10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(14, "svg", 6);
        ɵngcc0.ɵɵelementStart(15, "symbol", 11);
        ɵngcc0.ɵɵelement(16, "path", 12);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(17, "svg", 6);
        ɵngcc0.ɵɵelementStart(18, "symbol", 13);
        ɵngcc0.ɵɵelement(19, "path", 14);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("ngx_notification-msg-opened", ctx.isNotificationMsgOpened())("ngx_notification-msg-progress-bar", ctx.isProgressBarDisplayed());
        ɵngcc0.ɵɵproperty("ngStyle", ctx.getPosition());
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.isIconDisplayed());
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.header, " ");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", (ctx.msgs == null ? null : ctx.msgs.length) === 1);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", (ctx.msgs == null ? null : ctx.msgs.length) > 1);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.closeable);
    } }, directives: [ɵngcc1.NgStyle, ɵngcc1.NgIf, ɵngcc1.NgForOf], styles: ["[_ngcontent-%COMP%]:root{--ngx-notification-msg-delay:0;--ngx-notification-msg-color:none}.ngx_notification-msg[_ngcontent-%COMP%]{position:absolute;display:-webkit-box;display:flex;opacity:0;box-shadow:0 10px 19px 10px rgba(0,0,0,.04);border-radius:4px;background-color:#fff;font-family:Raleway,Arial,sans-serif;-webkit-transition:.2s cubic-bezier(.75,0,.75,.9);transition:.2s cubic-bezier(.75,0,.75,.9)}.ngx_notification-msg[_ngcontent-%COMP%]:hover::after{-webkit-animation-play-state:paused;animation-play-state:paused}.ngx_notification-msg-opened[_ngcontent-%COMP%]{opacity:1;-webkit-transition:250ms cubic-bezier(.2,0,.25,1);transition:250ms cubic-bezier(.2,0,.25,1)}.ngx_notification-msg-progress-bar[_ngcontent-%COMP%]::before{content:'';height:4px;width:100%;position:absolute;bottom:0;opacity:.3;border-bottom-left-radius:4px;border-bottom-right-radius:4px;background-color:var(--ngx-notification-msg-color)}.ngx_notification-msg-progress-bar[_ngcontent-%COMP%]::after{content:'';height:4px;position:absolute;bottom:0;border-bottom-left-radius:4px;background-color:var(--ngx-notification-msg-color);-webkit-animation:change-width var(--ngx-notification-msg-delay);animation:change-width var(--ngx-notification-msg-delay)}.ngx_notification-msg-icon-wrapper[_ngcontent-%COMP%]{width:81px;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;-webkit-box-pack:center;justify-content:center;justify-items:center;position:relative}.ngx_notification-msg-icon-wrapper[_ngcontent-%COMP%]   svg[_ngcontent-%COMP%]{height:24px;width:24px;color:var(--ngx-notification-msg-color);fill:currentColor;-webkit-transform:scale(1.167);transform:scale(1.167)}.ngx_notification-msg-icon-wrapper[_ngcontent-%COMP%]::after{content:'';height:calc(100% - 40px);width:1px;position:absolute;right:0;border-bottom-left-radius:4px;border-top-left-radius:4px;background-color:#f1f1f2}.ngx_notification-msg-content[_ngcontent-%COMP%]{padding:20px 30px}.ngx_notification-msg-header[_ngcontent-%COMP%]{line-height:1.5;font-size:15px;font-weight:700;color:#565154}.ngx_notification-msg-body[_ngcontent-%COMP%]{line-height:1.5;font-size:13px;color:#93908c}.ngx_notification-msg-body[_ngcontent-%COMP%]   ul[_ngcontent-%COMP%]{margin:8px 0;padding:0 0 0 30px}.ngx_notification-msg-body[_ngcontent-%COMP%]   ul[_ngcontent-%COMP%]:last-of-type{margin:8px 0 0}.ngx_notification-msg-close-button[_ngcontent-%COMP%]{height:16px;width:16px;position:absolute;top:8px;right:8px;outline:0;padding:0;font-size:11px;font-weight:600;background-color:transparent;cursor:pointer;box-sizing:content-box;border-radius:50%;border:0}.ngx_notification-msg-close-button[_ngcontent-%COMP%]   svg[_ngcontent-%COMP%]{color:#717075;fill:currentColor;height:inherit;width:inherit}.ngx_notification-msg-close-button[_ngcontent-%COMP%]:active{-webkit-animation:.1s close-click;animation:.1s close-click}@-webkit-keyframes change-width{0%{width:100%;border-bottom-right-radius:4px}100%{width:0;border-bottom-right-radius:0}}@keyframes change-width{0%{width:100%;border-bottom-right-radius:4px}100%{width:0;border-bottom-right-radius:0}}@-webkit-keyframes close-click{0%{border:0 solid #f1f1f2}100%{border:10px solid #f1f1f2;margin-top:-10px;margin-right:-10px}}@keyframes close-click{0%{border:0 solid #f1f1f2}100%{border:10px solid #f1f1f2;margin-top:-10px;margin-right:-10px}}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxNotificationMsgComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-notification-msg',
                template: "<div class=\"ngx_notification-msg\"\n     [class.ngx_notification-msg-opened]=\"isNotificationMsgOpened()\"\n     [class.ngx_notification-msg-progress-bar]=\"isProgressBarDisplayed()\"\n     [ngStyle]=\"getPosition()\"\n     (mouseenter)=\"mouseEnter()\"\n     (mouseleave)=\"mouseLeave()\">\n    <div class=\"ngx_notification-msg-icon-wrapper\"\n         *ngIf=\"isIconDisplayed()\">\n        <svg>\n            <use [attr.xlink:href]=\"statusToClass[status]\"/>\n        </svg>\n    </div>\n    <div class=\"ngx_notification-msg-content\">\n        <div class=\"ngx_notification-msg-header\">\n            {{header}}\n        </div>\n        <div class=\"ngx_notification-msg-body\"\n             *ngIf=\"msgs?.length === 1\">\n            {{msgs}}\n        </div>\n        <div class=\"ngx_notification-msg-body\"\n             *ngIf=\"msgs?.length > 1\">\n            <ul *ngFor=\"let msg of msgs\">\n                <li>{{msg}}</li>\n            </ul>\n        </div>\n        <button class=\"ngx_notification-msg-close-button\"\n                *ngIf=\"closeable\"\n                (click)=\"close()\">\n            <svg>\n                <use [attr.xlink:href]=\"'#clear'\"/>\n            </svg>\n        </button>\n    </div>\n\n    <svg display=\"none\">\n        <symbol width=\"24\" height=\"24\" viewBox=\"0 0 32 32\" id=\"00CC69\">\n            <path d=\"M16,0 C24.836556,0 32,7.163444 32,16 C32,24.836556 24.836556,32 16,32 C7.163444,32 0,24.836556 0,16 C0,7.163444 7.163444,0 16,0 Z M23.6,9.6 L13,20.2 L8.8,16 L7.4,17.4 L13,23 L25,11 L23.6,9.6 Z\"/>\n        </symbol>\n    </svg>\n\n    <svg display=\"none\">\n        <symbol width=\"24\" height=\"24\" viewBox=\"0 0 32 32\" id=\"FE355A\">\n            <path d=\"M16,0 C24.836556,0 32,7.163444 32,16 C32,24.836556 24.836556,32 16,32 C7.163444,32 0,24.836556 0,16 C0,7.163444 7.163444,0 16,0 Z M21.59,9 L16,14.59 L10.41,9 L9,10.41 L14.59,16 L9,21.59 L10.41,23 L16,17.41 L21.59,23 L23,21.59 L17.41,16 L23,10.41 L21.59,9 Z\"/>\n        </symbol>\n    </svg>\n\n    <svg display=\"none\">\n        <symbol width=\"24\" height=\"24\" viewBox=\"0 0 32 32\" id=\"0067FF\">\n            <path d=\"M16,0 C24.836556,0 32,7.163444 32,16 C32,24.836556 24.836556,32 16,32 C7.163444,32 0,24.836556 0,16 C0,7.163444 7.163444,0 16,0 Z M17,12 L15,12 L15,24 L17,24 L17,12 Z M17,8 L15,8 L15,10 L17,10 L17,8 Z\"/>\n        </symbol>\n    </svg>\n\n    <svg display=\"none\">\n        <symbol viewBox=\"0 0 24 24\" id=\"clear\">\n            <path d=\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\"/>\n        </symbol>\n    </svg>\n</div>\n",
                styles: [":root{--ngx-notification-msg-delay:0;--ngx-notification-msg-color:none}.ngx_notification-msg{position:absolute;display:-webkit-box;display:flex;opacity:0;box-shadow:0 10px 19px 10px rgba(0,0,0,.04);border-radius:4px;background-color:#fff;font-family:Raleway,Arial,sans-serif;-webkit-transition:.2s cubic-bezier(.75,0,.75,.9);transition:.2s cubic-bezier(.75,0,.75,.9)}.ngx_notification-msg:hover::after{-webkit-animation-play-state:paused;animation-play-state:paused}.ngx_notification-msg-opened{opacity:1;-webkit-transition:250ms cubic-bezier(.2,0,.25,1);transition:250ms cubic-bezier(.2,0,.25,1)}.ngx_notification-msg-progress-bar::before{content:'';height:4px;width:100%;position:absolute;bottom:0;opacity:.3;border-bottom-left-radius:4px;border-bottom-right-radius:4px;background-color:var(--ngx-notification-msg-color)}.ngx_notification-msg-progress-bar::after{content:'';height:4px;position:absolute;bottom:0;border-bottom-left-radius:4px;background-color:var(--ngx-notification-msg-color);-webkit-animation:change-width var(--ngx-notification-msg-delay);animation:change-width var(--ngx-notification-msg-delay)}.ngx_notification-msg-icon-wrapper{width:81px;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;-webkit-box-pack:center;justify-content:center;justify-items:center;position:relative}.ngx_notification-msg-icon-wrapper svg{height:24px;width:24px;color:var(--ngx-notification-msg-color);fill:currentColor;-webkit-transform:scale(1.167);transform:scale(1.167)}.ngx_notification-msg-icon-wrapper::after{content:'';height:calc(100% - 40px);width:1px;position:absolute;right:0;border-bottom-left-radius:4px;border-top-left-radius:4px;background-color:#f1f1f2}.ngx_notification-msg-content{padding:20px 30px}.ngx_notification-msg-header{line-height:1.5;font-size:15px;font-weight:700;color:#565154}.ngx_notification-msg-body{line-height:1.5;font-size:13px;color:#93908c}.ngx_notification-msg-body ul{margin:8px 0;padding:0 0 0 30px}.ngx_notification-msg-body ul:last-of-type{margin:8px 0 0}.ngx_notification-msg-close-button{height:16px;width:16px;position:absolute;top:8px;right:8px;outline:0;padding:0;font-size:11px;font-weight:600;background-color:transparent;cursor:pointer;box-sizing:content-box;border-radius:50%;border:0}.ngx_notification-msg-close-button svg{color:#717075;fill:currentColor;height:inherit;width:inherit}.ngx_notification-msg-close-button:active{-webkit-animation:.1s close-click;animation:.1s close-click}@-webkit-keyframes change-width{0%{width:100%;border-bottom-right-radius:4px}100%{width:0;border-bottom-right-radius:0}}@keyframes change-width{0%{width:100%;border-bottom-right-radius:4px}100%{width:0;border-bottom-right-radius:0}}@-webkit-keyframes close-click{0%{border:0 solid #f1f1f2}100%{border:10px solid #f1f1f2;margin-top:-10px;margin-right:-10px}}@keyframes close-click{0%{border:0 solid #f1f1f2}100%{border:10px solid #f1f1f2;margin-top:-10px;margin-right:-10px}}"]
            }]
    }], function () { return [{ type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc0.ElementRef }]; }, { status: [{
            type: Input
        }], direction: [{
            type: Input
        }], displayProgressBar: [{
            type: Input
        }], displayIcon: [{
            type: Input
        }], delay: [{
            type: Input
        }], closeable: [{
            type: Input
        }], destroy: [{
            type: Output
        }], messages: [{
            type: Input
        }], header: [{
            type: Input
        }], index: [{
            type: Input
        }] }); })();
    return NgxNotificationMsgComponent;
}());
export { NgxNotificationMsgComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxNotificationMsgComponent.DELAY_ON_CLICK;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.status;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.direction;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.displayProgressBar;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.displayIcon;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.header;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.delay;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.closeable;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.index;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.destroy;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.componentState;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.componentStates;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.statusToClass;
    /** @type {?} */
    NgxNotificationMsgComponent.prototype.msgs;
    /**
     * @type {?}
     * @private
     */
    NgxNotificationMsgComponent.prototype.NONE;
    /**
     * @type {?}
     * @private
     */
    NgxNotificationMsgComponent.prototype.closeTimeout;
    /**
     * @type {?}
     * @private
     */
    NgxNotificationMsgComponent.prototype.destroyTimeout;
    /**
     * @type {?}
     * @private
     */
    NgxNotificationMsgComponent.prototype.referencePointTimestamp;
    /**
     * @type {?}
     * @private
     */
    NgxNotificationMsgComponent.prototype.mouseEnterTimestamp;
    /**
     * @type {?}
     * @private
     */
    NgxNotificationMsgComponent.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    NgxNotificationMsgComponent.prototype.element;
}
/**
 * @record
 */
function INgxNotificationPosition() { }
if (false) {
    /** @type {?|undefined} */
    INgxNotificationPosition.prototype.top;
    /** @type {?|undefined} */
    INgxNotificationPosition.prototype.bottom;
    /** @type {?|undefined} */
    INgxNotificationPosition.prototype.left;
    /** @type {?|undefined} */
    INgxNotificationPosition.prototype.right;
    /** @type {?} */
    INgxNotificationPosition.prototype.transform;
}
/** @enum {string} */
var NgxNotificationMsgComponentState = {
    CLOSE: 'CLOSE',
    OPEN: 'OPEN',
};
export { NgxNotificationMsgComponentState };
/** @enum {string} */
var NgxNotificationStatusMsg = {
    SUCCESS: 'SUCCESS',
    FAILURE: 'FAILURE',
    INFO: 'INFO',
    NONE: 'NONE',
};
export { NgxNotificationStatusMsg };
/** @enum {string} */
var NgxNotificationDirection = {
    TOP: 'TOP',
    TOP_RIGHT: 'TOP_RIGHT',
    TOP_LEFT: 'TOP_LEFT',
    BOTTOM: 'BOTTOM',
    BOTTOM_RIGHT: 'BOTTOM_RIGHT',
    BOTTOM_LEFT: 'BOTTOM_LEFT',
};
export { NgxNotificationDirection };
/**
 * @record
 */
export function INgxNotificationMsgConfig() { }
if (false) {
    /** @type {?|undefined} */
    INgxNotificationMsgConfig.prototype.status;
    /** @type {?|undefined} */
    INgxNotificationMsgConfig.prototype.direction;
    /** @type {?|undefined} */
    INgxNotificationMsgConfig.prototype.header;
    /** @type {?} */
    INgxNotificationMsgConfig.prototype.messages;
    /** @type {?|undefined} */
    INgxNotificationMsgConfig.prototype.delay;
    /** @type {?|undefined} */
    INgxNotificationMsgConfig.prototype.displayIcon;
    /** @type {?|undefined} */
    INgxNotificationMsgConfig.prototype.displayProgressBar;
    /** @type {?|undefined} */
    INgxNotificationMsgConfig.prototype.closeable;
}
