// ngx-notification-msg/esm5/lib/ngx-notification-msg.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgxNotificationMsgComponent } from './ngx-notification-msg.component';
import { CommonModule } from '@angular/common';
import * as ɵngcc0 from '@angular/core';
var NgxNotificationMsgModule = /** @class */ (function () {
    function NgxNotificationMsgModule() {
    }
NgxNotificationMsgModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxNotificationMsgModule });
NgxNotificationMsgModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxNotificationMsgModule_Factory(t) { return new (t || NgxNotificationMsgModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxNotificationMsgModule, { declarations: function () { return [NgxNotificationMsgComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgxNotificationMsgComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxNotificationMsgModule, [{
        type: NgModule,
        args: [{
                entryComponents: [NgxNotificationMsgComponent],
                declarations: [NgxNotificationMsgComponent],
                imports: [CommonModule],
                exports: [NgxNotificationMsgComponent]
            }]
    }], function () { return []; }, null); })();
    return NgxNotificationMsgModule;
}());
export { NgxNotificationMsgModule };
