// simplebar-angular/esm5/lib/simplebar-angular.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { SimplebarAngularComponent } from './simplebar-angular.component';
import * as ɵngcc0 from '@angular/core';
var SimplebarAngularModule = /** @class */ (function () {
    function SimplebarAngularModule() {
    }
SimplebarAngularModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: SimplebarAngularModule });
SimplebarAngularModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function SimplebarAngularModule_Factory(t) { return new (t || SimplebarAngularModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SimplebarAngularModule, { declarations: function () { return [SimplebarAngularComponent]; }, exports: function () { return [SimplebarAngularComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SimplebarAngularModule, [{
        type: NgModule,
        args: [{
                declarations: [SimplebarAngularComponent],
                imports: [],
                exports: [SimplebarAngularComponent]
            }]
    }], function () { return []; }, null); })();
    return SimplebarAngularModule;
}());
export { SimplebarAngularModule };
