// ng-wizard/esm5/lib/core/ng-wizard-step-content.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, ViewContainerRef } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var NgWizardStepContentDirective = /** @class */ (function () {
    function NgWizardStepContentDirective(viewContainerRef) {
        this.viewContainerRef = viewContainerRef;
    }
    NgWizardStepContentDirective.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    NgWizardStepContentDirective = __decorate([ __metadata("design:paramtypes", [ViewContainerRef])
    ], NgWizardStepContentDirective);
NgWizardStepContentDirective.ɵfac = function NgWizardStepContentDirective_Factory(t) { return new (t || NgWizardStepContentDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef)); };
NgWizardStepContentDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: NgWizardStepContentDirective, selectors: [["", "ngWizardStepContent", ""]] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgWizardStepContentDirective, [{
        type: Directive,
        args: [{
                selector: '[ngWizardStepContent]'
            }]
    }], function () { return [{ type: ɵngcc0.ViewContainerRef }]; }, null); })();
    return NgWizardStepContentDirective;
}());
export { NgWizardStepContentDirective };
