// ng-wizard/esm5/lib/core/ng-wizard.service.js
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { NgWizardDataService } from './ng-wizard-data.service';
import * as i0 from "@angular/core";
import * as i1 from "./ng-wizard-data.service";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './ng-wizard-data.service';
var NgWizardService = /** @class */ (function () {
    function NgWizardService(ngWizardDataService) {
        this.ngWizardDataService = ngWizardDataService;
    }
    NgWizardService.prototype.reset = function () {
        this.ngWizardDataService.resetWizard();
    };
    NgWizardService.prototype.next = function () {
        this.ngWizardDataService.showNextStep();
    };
    NgWizardService.prototype.previous = function () {
        this.ngWizardDataService.showPreviousStep();
    };
    NgWizardService.prototype.show = function (index) {
        this.ngWizardDataService.showStep(index);
    };
    NgWizardService.prototype.theme = function (theme) {
        this.ngWizardDataService.setTheme(theme);
    };
    NgWizardService.prototype.stepChanged = function () {
        return this.ngWizardDataService.stepChangedArgs$;
    };
    NgWizardService.ctorParameters = function () { return [
        { type: NgWizardDataService }
    ]; };
    NgWizardService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgWizardService_Factory() { return new NgWizardService(i0.ɵɵinject(i1.NgWizardDataService)); }, token: NgWizardService, providedIn: "root" });
    NgWizardService = __decorate([ __metadata("design:paramtypes", [NgWizardDataService])
    ], NgWizardService);
NgWizardService.ɵfac = function NgWizardService_Factory(t) { return new (t || NgWizardService)(ɵngcc0.ɵɵinject(ɵngcc1.NgWizardDataService)); };
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgWizardService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: ɵngcc1.NgWizardDataService }]; }, null); })();
    return NgWizardService;
}());
export { NgWizardService };
