// ngx-print/esm5/lib/ngx-print.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgxPrintDirective } from './ngx-print.directive';
import * as ɵngcc0 from '@angular/core';
var NgxPrintModule = /** @class */ (function () {
    function NgxPrintModule() {
    }
NgxPrintModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxPrintModule });
NgxPrintModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxPrintModule_Factory(t) { return new (t || NgxPrintModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxPrintModule, { declarations: function () { return [NgxPrintDirective]; }, exports: function () { return [NgxPrintDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxPrintModule, [{
        type: NgModule,
        args: [{
                declarations: [NgxPrintDirective],
                imports: [],
                exports: [NgxPrintDirective]
            }]
    }], function () { return []; }, null); })();
    return NgxPrintModule;
}());
export { NgxPrintModule };
